src/surrogate_heat_driver.cpp

```diff
@@ -3,54 +3,131 @@
   clad_inner_radius_ = node.child("clad_inner_radius").text().as_double();
   clad_outer_radius_ = node.child("clad_outer_radius").text().as_double();
   pellet_radius_ = node.child("pellet_radius").text().as_double();
   n_fuel_rings_ = node.child("fuel_rings").text().as_int();
   n_clad_rings_ = node.child("clad_rings").text().as_int();
+  n_pins_x_ = node.child("n_pins_x").text().as_int();
+  n_pins_y_ = node.child("n_pins_y").text().as_int();
+  n_pins_ = n_pins_x_ * n_pins_y_;
+  pin_pitch_ = node.child("pin_pitch").text().as_double();
+
+  // Determine thermal-hydraulic parameters for fluid phase
+  inlet_temperature_ = node.child("inlet_temperature").text().as_double();
+  mass_flowrate_ = node.child("mass_flowrate").text().as_double();
+  n_channels_ = (n_pins_x_ + 1) * (n_pins_y_ + 1);
 
+  // Determine solver parameters
+  if (node.child("max_subchannel_its"))
+    max_subchannel_its_ = node.child("max_subchannel_its").text().as_int();
+  if (node.child("subchannel_tol_h"))
+    subchannel_tol_h_= node.child("subchannel_tol_h").text().as_double();
+  if (node.child("subchannel_tol_p"))
+    subchannel_tol_p_= node.child("subchannel_tol_p").text().as_double();
+  if (node.child("heat_tol"))
+    heat_tol_ = node.child("heat_tol").text().as_double();
+
+  // check validity of user input
   Expects(clad_inner_radius_ > 0);
   Expects(clad_outer_radius_ > clad_inner_radius_);
   Expects(pellet_radius_ < clad_inner_radius_);
   Expects(n_fuel_rings_ > 0);
   Expects(n_clad_rings_ > 0);
+  Expects(n_pins_x_ > 0);
+  Expects(n_pins_y_ > 0);
+  Expects(pin_pitch_ > 2.0 * clad_outer_radius_);
+  Expects(mass_flowrate_ > 0.0);
+  Expects(inlet_temperature_ > 0.0);
+  Expects(max_subchannel_its_ > 0);
+  Expects(subchannel_tol_h_ > 0.0);
+  Expects(subchannel_tol_p_ > 0.0);
+  Expects(heat_tol_ > 0.0);
 
-  // Get pin locations
-  // TODO: Switch to get_node_xarray on OpenMC update
-  auto pin_locations = openmc::get_node_array<double>(node, "pin_centers");
-  if (pin_locations.size() % 2 != 0) {
-    throw std::runtime_error{"Length of <pin_centers> must be a multiple of two"};
+  // Set pin locations, where the center of the assembly is assumed to occur at
+  // x = 0, y = 0. It is also assumed that the rod-boundary separation in the
+  // x and y directions is the same and equal to half the pitch.
+  double assembly_width_x = n_pins_x_ * pin_pitch_;
+  double assembly_width_y = n_pins_y_ * pin_pitch_;
+
+  pin_centers_.resize({n_pins_});
+  for (int row = 0; row < n_pins_y_; ++row) {
+    for (int col = 0; col < n_pins_x_; ++col) {
+      int pin_index = row * n_pins_x_ + col;
+      pin_centers_(pin_index, 0) = -assembly_width_x / 2.0 + pin_pitch_ / 2.0 + col * pin_pitch_;
+      pin_centers_(pin_index, 1) = assembly_width_y / 2.0 - (pin_pitch_ / 2.0 + row * pin_pitch_);
+    }
+  }
+
+  // Initialize the channels
+  ChannelFactory channel_factory(pin_pitch_, clad_outer_radius_);
+
+  for (int row = 0; row < n_pins_y_ + 1; ++row) {
```

Comment:
I've started using [gsl::index](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-subscripts) for loop indices if you want to do the same.